Improving Cypher and BOLT interaction

pontusmelke opened the pull request:
Initially this was supposed to be a small change to improve the interaction between Cypher and Bolt but it sort of grew from there. Sorry about that.

Main idea is that Bolt now gets `Values` from Cypher instead of raw `Objects`, which allows for more efficient serialization without resorting to a big `instance of` switch. 

In order to achieve this the following changes were made:
- Interpreted runtime is now a `Map[String, AnyValue]` instead of a `Map[String, Any]`
- All expressions in the interpreted runtime now returns `AnyValue` instead of `Any`
- The compiled runtime now _boxes_ primitives to `AnyValue` instead of boxing to `Object` before interacting with the visitor.
- Some levels of indirections in the final `Result` has been removed.
- the `values` module has been updated in order to

NOTE: The failing docs build is due to internal changes in this PR. https://github.com/neo4j/neo4j-documentation/pull/113 will addresses these failures, so we should merge these PRs at more or less the same time.

pontusmelke committed:
Let bolt depend on value module

pontusmelke committed:
Allow lists to contain NO_VALUE

pontusmelke committed:
Add AnyValues.of

Add factory method for creating all of the Cypher types, mostly to be
used in testing.

pontusmelke committed:
Added a packer that handles Values

Implemented `Packer` as a `AnyValueWriter`

pontusmelke committed:
Refactor how execution results are handled in cypher

pontusmelke committed:
Updated enterprise cypher module after refactoring execution result

pontusmelke committed:
Fixed bugs in displaying Notification

pontusmelke committed:
Added several list implementations

In order to facilitate cypher implementation we add multiple
list variants.

pontusmelke committed:
Added utilities to Values

Added a bunch of minor help methods in order to facilitate cypher
integration

pontusmelke committed:
Changed to a type runtime

ExecutionContext is now a `Map[String, AnyValue]` and all expressions now
produce `AnyValue`.

pontusmelke committed:
Introduced wrappers for node proxy and rel proxy

For handling some of the legacy we introduce a wrapped lazy loading
node value and edge value.

pontusmelke committed:
Fixed errors from rebase

pontusmelke committed:
Fix how procedures handle input

pontusmelke committed:
Fix how parameters are handled

Parameters are turned into a `Map[String, AnyValue]` before
calling the underlying layer. Preferably this should be done even
earlier but compatability makes it tricky.

pontusmelke committed:
Update tests

Update a lot of the test to use `AnyValue` instead of `Any`

pontusmelke committed:
Fix rebase errors

pontusmelke committed:
Fix tests

pontusmelke committed:
Added typed accept method to Result

In order to be able to use a typed Value directly we add a typed version
of the `accept` method to InternalExecutionResult

pontusmelke committed:
Updated compiled runtime to use typed accept

pontusmelke committed:
Fix rebase errors

pontusmelke committed:
Update bolt to use internal accept method

pontusmelke committed:
Refactor bolt module

- Remove unused stuff
- Make packing/unpacking symmetric in types

pontusmelke committed:
Fix rebase errors

pontusmelke committed:
Fix tests

pontusmelke committed:
Values are immutable, compute hash once

Since values are immutable we only need to compute the hash value once
and memoize the results. Note that ints are guaranteed to be read
atomically so it is safe to use non-volatile read, see `java.util.String`.

pontusmelke committed:
Don't swallow NotFoundException

pontusmelke committed:
Update BOLT integration tests

pontusmelke committed:
computeHash should only be renamed in values

pontusmelke committed:
Update user-defined aggregation functions

User-defined aggregation functions wasn't using `Values` properly

pontusmelke committed:
Update security testing infrastructure

pontusmelke committed:
Fix rebase errors

pontusmelke committed:
Fixed rebase errors

pontusmelke committed:
Reduce the number of proxies created

When going from Values to Object we shortcut the serializer in order to
avoid having multiple proxies being created.

pontusmelke committed:
Fix failing test.

Test was failing due to insufficient mocking

pontusmelke committed:
Update procedure interaction tests

The `ProcedureInteractionTestBase` needed update in order to handle
Values for Bolt tests.

pontusmelke committed:
Handle results from standalone procedure calls

pontusmelke committed:
Update authentication tests to handle Values

pontusmelke committed:
GetServers had the wrong signature

The LegacyGetServersProcedure presented itself with the wrong signature
which caused issues in the updated runtime.

pontusmelke committed:
Fix rebase errors

pontusmelke committed:
Wrong size of result array

The size of the field array for standalone procedure results was wrong.

pontusmelke committed:
Handle Value in RelationshipEndPoints

pontusmelke committed:
Fixed string output for constraints

pontusmelke committed:
Fix bad name in black list

pontusmelke committed:
Fix bug in how blacklisting works

Blacklists where not considering control queries correctly which made it
in impossible to blacklist tests that contains control queries.

pontusmelke committed:
Blacklist tests

There is an error in the tck that enforces a strict ordering in `keys`,
until the tck is updated tests that are struck by this will be blacklisted.

pontusmelke committed:
Update error definitions in test suite

Error message has changed slightly and was no longer being matched in the test suite.

pontusmelke committed:
Fixed failing test

craigtaverner commented on community/bolt/src/test/java/org/neo4j/bolt/v1/transport/integration/AuthenticationIT.java:
```diff
@@ -564,4 +566,9 @@ public boolean gotSpecialMessage()
         }
     }
+
+    public static Map<String,AnyValue> mapValue( Object... objects )
```
This method is never used. Perhaps remove it again?

craigtaverner commented on community/bolt/src/test/java/org/neo4j/bolt/v1/transport/integration/TransportSessionIT.java:
```diff
@@ -41,6 +41,8 @@
 import org.neo4j.graphdb.factory.GraphDatabaseSettings;
 import org.neo4j.helpers.HostnamePort;
 import org.neo4j.kernel.api.exceptions.Status;
+import org.neo4j.values.storable.Values;
```
Unused imports

craigtaverner commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/javacompat/ResultRecord.java:
```diff
@@ -17,25 +17,36 @@
-    Path getPath( String key );
+    private final AnyValue[] fields;
+
+    public ResultRecord( int size )
```
Constructor never used.

craigtaverner commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/javacompat/ResultRowImpl.java:
```diff
@@ -37,7 +37,8 @@ public ResultRowImpl( Map<String, Object> results )
 
-    public ResultRowImpl( )
+    public
+    ResultRowImpl( )
```
Formatting?

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v2_3/Compatibility.scala:
```diff
@@ -61,25 +61,29 @@ trait Compatibility {
-        notificationLogger,
-        preParsedQuery.planner.name,
-        Some(as2_3(preParsedQuery.offset)), tracer))
+                                preParsedQuery.rawStatement,
```
Most of the classes changes are reformatting. Makes it hard to see the real change, as well as hard to forward merge fixes.

pontusmelke replied:
tried to revert most of the reformatting changes

fickludd commented on community/values/pom.xml:
```diff
@@ -34,36 +35,41 @@
     <dependency>
+      <groupId>org.neo4j</groupId>
+      <artifactId>neo4j-graphdb-api</artifactId>
+      <version>${project.version}</version>
```
Could we remove this dep. by moving the converters somewhere else. This feels a non-intuitive to me. On the other hand, I guess `neo4j-graphdb-api` is very stable and doesn't drag in much else, so maybe it's fine.

pontusmelke replied:
Lets talk about it and move it in a separate PR?

fickludd commented on community/values/src/main/java/org/neo4j/values/BaseToObjectValueWriter.java:
```diff
@@ -0,0 +1,545 @@
+    public void writePath( NodeValue[] nodes, EdgeValue[] edges ) throws RuntimeException
+    {
+        assert nodes != null;
+        assert nodes.length > 0;
```
´assert edges.length == nodes.length - 1´?

pontusmelke replied:
fixed

fickludd commented on community/values/src/main/java/org/neo4j/values/storable/ArrayValue.java:
```diff
@@ -20,12 +20,15 @@
 package org.neo4j.values.storable;
 
 import org.neo4j.values.SequenceValue;
+import org.neo4j.values.AnyValue;
```
not used

fickludd commented on community/values/src/main/java/org/neo4j/values/BaseToObjectValueWriter.java:
```diff
@@ -0,0 +1,545 @@
+
+import static org.neo4j.helpers.collection.Iterators.iteratorsEqual;
+
+public abstract class BaseToObjectValueWriter<E extends Exception> implements AnyValueWriter<E>
```
This class could use some javadoc. It took me a while to understand how it works and why we need it. Would still suggest we move `graphdb-api` related stuff out of this module, but let's to that in a separate PR in that case.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/Add.scala:
```diff
@@ -20,27 +20,31 @@
+      case (x: IntegralValue, y: IntegralValue) => Values.longValue(StrictMath.addExact(x.longValue(),y.longValue()))
+      case (x: NumberValue, y: NumberValue) => Values.doubleValue(x.doubleValue() + y.doubleValue())
+      case (x: TextValue, y: TextValue) => Values.stringValue(x.stringValue() + y.stringValue())
+      case (IsList(x),  IsList(y)) => VirtualValues.concat(x, y)
```
Extra space.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/Add.scala:
```diff
@@ -20,27 +20,31 @@
 import org.neo4j.cypher.internal.frontend.v3_3.symbols._
+import org.neo4j.values._
+import org.neo4j.values.storable._
+import org.neo4j.values.virtual.{ListValue, VirtualValues}
```
unused import

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/helpers/MapSupport.scala:
```diff
@@ -19,57 +19,33 @@
+
+  def castToMap: PartialFunction[AnyValue, MapValue] = {
+    case x: MapValue => x
+    case x: NodeValue  => x.properties()
```
In the older version we made clear use of the QueryContext to access the kernel to lookup properties. This new approach implies that either the properties are contained in the node (eagerly looked up), or the Node has capability to do this lookup, much like the NodeProxy which we wanted to get away from. I like that the code is so much cleaner, I just wonder if we've moving away from the attempt to control kernel access that we had before?

fickludd commented on community/values/src/main/java/org/neo4j/values/storable/FloatingPointArray.java:
```diff
@@ -19,6 +19,8 @@
  */
 package org.neo4j.values.storable;
 
+import org.neo4j.values.AnyValue;
```
not used

fickludd commented on community/values/src/main/java/org/neo4j/values/storable/IntegralArray.java:
```diff
@@ -19,6 +19,8 @@
  */
 package org.neo4j.values.storable;
 
+import org.neo4j.values.AnyValue;
```
not used

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/DistanceFunction.scala:
```diff
@@ -26,19 +26,20 @@ import org.neo4j.cypher.internal.compatibility.v3_3.runtime.pipes.QueryState
-      case (geometry1: Point, geometry2: Point) => calculateDistance(geometry1, geometry2)
-      case (geometry1: Geometry, geometry2: Geometry) => calculateDistance(geometry1, geometry2)
+      case (x, y) if x == Values.NO_VALUE || y == Values.NO_VALUE => Values.NO_VALUE
+      case (geometry1: PointValue, geometry2: PointValue) => calculateDistance(geometry1, geometry2)
```
We have lost support for Geometry conversion to Points. The coerceToPoint method still exists, but is no longer used.

pontusmelke replied:
No we still do that the `calculateDistance` function that takes in geometries handles that. I'll try to remove the unused method though.

pontusmelke replied:
Or to clarify, we will coerce geometries to points earlier, whenever we try to put a geometry into the context we convert it to a point since that is the only geometry we can handle

craigtaverner replied:
So you suggest we bring back the conversion later when we support Geometry again? That is possible, I just thought we had tests that would fail. Either those tests have been removed, or I'm remembering wrong.

pontusmelke replied:
we do support geometry but we only have points so far. If a user pass in a geometry via parameters etc we will try to convert it to a point since that is the only geometry we know about. Same as before but we only convert/fail earlier.

fickludd commented on community/values/src/test/java/org/neo4j/values/virtual/ListTest.java:
```diff
@@ -198,4 +202,37 @@ public void shouldRecurseAndCoerce()
+                    VirtualValues.take( list( 1L, 4L, 7L, 10L, 13L ), 3 ),
+                    VirtualValues.transform( list( 0L, 3L, 6L ),
+                            anyValue -> Values.longValue( ((LongValue) anyValue).longValue() + 1L ) ),
+                    VirtualValues.reverse( list( 7L, 4L, 1L ) )};
```
I really like that you've added these ListValue manipulator methods. Did you consider putting them on the `ListValue` class instead?

pontusmelke replied:
yes, we have taken the static approach for other stuff so figured I'd stick with that. Maybe we can discuss this and change in a future PR if that makes more sense.

fickludd commented on community/values/src/test/java/org/neo4j/values/virtual/ListTest.java:
```diff
@@ -198,4 +202,37 @@ public void shouldRecurseAndCoerce()
+                assertArrayEquals(
+                        format( "%s.asArray != %s.toArray", list1.getClass().getSimpleName(),
+                                list2.getClass().getSimpleName() ),
+                        list1.asArray(), list2.asArray() );
```
Maybe adding a similar set of lists that are *not* equal would be good as a negative test?

fickludd replied:
And also, we should add a hash code check here, to be certain of consistent map behaviour.

pontusmelke replied:
`assertEquals` called on the line above already checks that `a == b && b ==a && a.hashCode == b.hashCode`

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/FilterFunction.scala:
```diff
@@ -21,15 +21,22 @@ package org.neo4j.cypher.internal.compatibility.v3_3.runtime.commands.expression
+
+  def compute(value: AnyValue, m: ExecutionContext)(implicit state: QueryState) = {
+    val traversable = makeTraversable(value)
+    VirtualValues.filter(traversable, new java.util.function.Function[AnyValue, java.lang.Boolean]() {
```
Would it be possible to have a lambda here for cleanliness? Pehaps not possible on the scala/java interface, I guess. Or we have a scala version of VirtualValues.filter that wraps the java version.

pontusmelke replied:
Yeah the main target is java so here scala will have to suffer a bit. I think when we upgrade to newer version of scala it will handle java lambdas better. We could of course add some utility but that will also add indirection, I think I prefer faster over prettier in the runtime.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/IndexedInclusiveLongRange.scala:
```diff
@@ -43,6 +43,7 @@ case class IndexedInclusiveLongRange(start: Long, end: Long, step: Long) extends
     }
   }
 
+  hashCode()
```
What does this do?

pontusmelke replied:
it computes the `hashCode` at a random place in the code. It has now been removed.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/PointFunction.scala:
```diff
@@ -19,24 +19,31 @@
     case x => throw new CypherTypeException(s"Expected a map but got $x")
   }
 
+  private def containsNull(map: MapValue) = {
```
Is there perhaps value in having this method on MapValue itself? It looks a bit ugly in scala code.

pontusmelke replied:
maybe

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/ProjectedPath.scala:
```diff
@@ -21,7 +21,9 @@ package org.neo4j.cypher.internal.compatibility.v3_3.runtime.commands.expression
-import org.neo4j.graphdb.{Node, Relationship}
+import org.neo4j.values.AnyValue
+import org.neo4j.values.storable.Values
+import org.neo4j.values.virtual.{EdgeValue, ListValue, NodeValue}
```
Unused imports

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/RelationshipEndPoints.scala:
```diff
@@ -20,21 +20,19 @@
-      else
-        state.query.relationshipEndNode(rel)
+      val rel = castOrFail[EdgeValue](value)
+      if (start) rel.startNode()
```
As in my comment with MapUtils above, we are no longer using the QueryContext to access kernel resources, and instead using the RelationshipProxy. At the very least this would hide dbHits in the plan profile, but it also feels like we are not moving away from the use of RelationshipProxy in Cypher code.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/ShortestPathExpression.scala:
```diff
@@ -136,7 +153,8 @@ case class ShortestPathExpression(shortestPathPattern: ShortestPath, predicates:
                                        predicate: Predicate, relName: String,
                                        currentNodePredicates: Seq[KernelPredicate[PropertyContainer]])
-                                      (implicit state: QueryState): (Expander, Seq[KernelPredicate[PropertyContainer]]) = {
+                                      (implicit
```
Line break in odd place.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/predicates/ComparablePredicate.scala:
```diff
@@ -55,29 +63,31 @@ abstract sealed class ComparablePredicate(val left: Expression, val right: Expre
-      case (_, null) => None
-      case _         => Some(Equivalent(a1) == b1)
+      case (x, y) if x == Values.NO_VALUE || y == Values.NO_VALUE => None
+      case (x: ArrayValue, y: ListValue) => Some(VirtualValues.fromArray(x).equals(y))
```
I wonder if it would be nicer to have ListValue and ArrayValue be able to handle equals of each others type? Here you convert always to ListValue. Is that most efficient?

pontusmelke replied:
I think it does now, I think someone fixed that while I was on vacation. Good catch

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/predicates/ComparablePredicate.scala:
```diff
@@ -55,29 +63,31 @@ abstract sealed class ComparablePredicate(val left: Expression, val right: Expre
     (a1, b1) match {
-      case (null, _) => None
-      case (_, null) => None
-      case _         => Some(Equivalent(a1) == b1)
```
Equivalent seemed to deal with quite a few things, including float/double comparisons and eagerisation of collections. I guess we have all the same capabilities in the values module now.

pontusmelke replied:
yes

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/predicates/Predicate.scala:
```diff
@@ -167,52 +195,64 @@ case class PropertyExists(variable: Expression, propertyKey: KeyToken) extends P
-    case lit: Literal => LiteralRegularExpression(lhsExpr.rewrite(f), lit)(converter)
-    case other        => RegularExpression(lhsExpr.rewrite(f), other)(converter)
-  })
+                                                   case lit: Literal => LiteralRegularExpression(lhsExpr.rewrite(f),
```
This file is full of reformating changes, some of which (like this one) are quite ugly. Did we need to do this much reformatting?

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/VirtualValues.java:
```diff
@@ -39,12 +49,103 @@ private VirtualValues()
+        int startPoint = 0;
+        for ( ListValue list : lists )
+        {
+            System.arraycopy( list.asArray(), 0, anyValues, startPoint, list.size() );
```
These `asArray` calls might allocate extra transient arrays depending on the underlying list implementation. Having a `ListValue.copyToArray( AnyValue[] target, int startPoint )` method instead could solve this (here and methods below).

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/NodeValue.java:
```diff
@@ -68,4 +63,107 @@ public String toString()
+            TextArray l = labels;
+            if ( l == null )
+            {
+                synchronized ( this )
```
This is the only place where I see concurrency-safe lazy vals, e.g. `ListValue` and `AnyValue` hash do not do this. I guess this was fuelled by some need?

pontusmelke replied:
My thinking is that we do create a lot of `Node` and `Relationship` - and hence now `NodeValue` and `EdgeValue` - in the interpreted runtime. Most of these object are not actually returned to the user and thus eagerly pulling in labels, types and properties eagerly would add a big cost for all these intermediate objects. In this way the labels and properties will only be fetched when bolt actually tries to serialize the objects.

In a new glorious runtime I think we should be a lot more careful about creating these intermediate objects, but I felt addressing all the cases where we create nodes and relationships unnecessarily in this PR would of had serious impact on my motivation for finishing this piece of work :)

fickludd replied:
Aha, yes I see. What I actually was wondering about was the *concurrency-safiness*?

pontusmelke replied:
NodeValues can be handled from multiple threads

fickludd commented on community/values/src/test/java/org/neo4j/values/storable/ValueComparisonTest.java:
```diff
@@ -87,7 +84,10 @@
+            Double.NaN,
+
+            // OTHER
+            null
```
Moving NO_VALUE to the end will change behavior from `3.2` in the kernel at least. Does that matter? I don't know. Does cypher 3.2 put nulls last?

pontusmelke replied:
Yes, `null` is last, https://github.com/neo4j/neo4j/blob/3.3/community/cypher/cypher-compiler-3.3/src/main/java/org/neo4j/cypher/internal/compiler/v3_3/common/CypherOrderability.java#L131

fickludd commented on community/values/src/main/java/org/neo4j/values/storable/Values.java:
```diff
@@ -42,8 +42,15 @@
+    public static final BooleanValue FALSE = Values.booleanValue( false );
+    public static final TextValue EMPTY_STRING = Values.stringValue( "" );
+    public static final DoubleValue E = Values.doubleValue( Math.E );
+    public static final DoubleValue PI = Values.doubleValue( Math.PI );
```
Very nice!

fickludd commented on community/values/src/main/java/org/neo4j/values/storable/Values.java:
```diff
@@ -143,87 +150,87 @@ public static Value numberValue( Number number )
     }
 
-    public static Value longValue( long value )
+    public static LongValue longValue( long value )
```
And been wanting to return specific types too!

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/VirtualValues.java:
```diff
@@ -19,18 +19,28 @@
 public final class VirtualValues
 {
+    public static MapValue EMPTY_MAP = new MapValue( Collections.emptyMap() );
+    public static ListValue EMPTY_LIST = new ListValue.ArrayListValue( new AnyValue[0] );
```
These guys should be `final`. Good to have around.

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/VirtualValues.java:
```diff
@@ -39,12 +49,103 @@ private VirtualValues()
+
+    public static ListValue reverse( ListValue list )
+    {
+        return new ListValue.ReversedList( list );
```
I like all these functional methods, but we to make a few unit tests to verify such central functionality.

pontusmelke replied:
added unit tests

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/VirtualValues.java:
```diff
@@ -80,26 +184,44 @@ public static EdgeReference edge( long id )
     {
+        assert nodes != null;
+        assert edges != null;
+        if ( nodes.length + edges.length % 2 == 0 )
```
you want `if ( ( nodes.length + edges.length ) % 2 == 0 )` or maybe `nodes.length != edges.length + 1`?

fickludd commented on community/values/src/main/java/org/neo4j/values/result/QueryResult.java:
```diff
@@ -17,18 +17,38 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.bolt.v1.messaging.infrastructure;
+package org.neo4j.values.result;
```
This move is a little bit mystic to me. I'm not saying it's bad, I would just like to understand the reason, as the interface it not clearly values related. Also some javadoc would be nice.

pontusmelke replied:
Yes I agree. It is a little bit values related since the visitor uses values. My grand plan was to have the result being defined outside cypher but maybe it should live in its own module. I don't know, lets discuss and move in a separate PR?

fickludd reviewed (changes requested):
I have been reviewing all changes in the values module, and nothing else.

This is really great work! Impressive work on specializing implementations and general improvements all around. Like the functional style methods on `ListValue`.

Been doing some nitpicking here and there, it's a big PR so there were things to say. The only crucial parts are some cases where I would like some more tests. 

Also, I'm completely fine with merging this PR and addressing comments on the values in a separate PR, to chunk the work up and avoid rebase issues and all that.

pontusmelke committed:
Removed unused methods and imports + formatting

pontusmelke committed:
Reintroduce context to IsMap

Although it was convenient it wasn't the correct thing to do. Now all
operations when considering a node or relationship as a map are routed via
the QueryContext

pontusmelke committed:
Lists and arrays can now be compared natively

No need to do an extra check, arrays and values can now be compared in
the values module.

pontusmelke committed:
Use context for finding opposing node

pontusmelke committed:
JavaDoc for BaseToObjectValueWriter

pontusmelke committed:
constants should be final

pontusmelke committed:
Changed how concatenation of ListValue

Instead of allocating array we add a ConcatListValue that doesn't do any
unnecessary allocation.

pontusmelke committed:
Unit test for ranges

pontusmelke committed:
Unit test for filtered lists

pontusmelke committed:
Unit test for list slices

pontusmelke committed:
Unit test for list transformation

pontusmelke committed:
Unit test for reversing lists

pontusmelke committed:
Added negative test for list equalities

pontusmelke committed:
Put back constructor used by generated code

pontusmelke committed:
Fixed signature of v2 getServers procedure
